hydra-node/src/Hydra/Chain/Direct/Tx.hs

```diff
@@ -217,26 +241,51 @@ commitTx ::
   ScriptRegistry ->
   HeadId ->
   Party ->
-  -- | The UTxO to commit to the Head along with witnesses.
-  UTxO' (TxOut CtxUTxO, Witness WitCtxTxIn) ->
+  -- | The UTxO to commit to the Head.
+  UTxO ->
+  -- | _Blueprint_ transaction.
+  Tx ->
   -- | The initial output (sent to each party) which should contain the PT and is
   -- locked by initial script
   (TxIn, TxOut CtxUTxO, Hash PaymentKey) ->
   Tx
-commitTx networkId scriptRegistry headId party utxoToCommitWitnessed (initialInput, out, vkh) =
-  unsafeBuildTransaction $
-    emptyTxBody
-      & addInputs [(initialInput, initialWitness)]
-      & addReferenceInputs [initialScriptRef]
-      & addInputs committedTxIns
-      & addExtraRequiredSigners [vkh]
-      & addOutputs [commitOutput]
-      & setTxMetadata (TxMetadataInEra $ mkHydraHeadV1TxName "CommitTx")
+commitTx networkId scriptRegistry headId party utxoToCommit blueprintTx (initialInput, out, vkh) =
+  let
+    ledgerBlueprintTx =
+      toLedgerTx blueprintTx
+        & bodyTxL . inputsTxBodyL <>~ Set.singleton (toLedgerTxIn initialInput)
+        & bodyTxL . inputsTxBodyL <>~ Set.fromList (toLedgerTxIn <$> committedTxIns)
+        & bodyTxL . referenceInputsTxBodyL <>~ Set.fromList [toLedgerTxIn initialScriptRef]
+        & bodyTxL . outputsTxBodyL %~ (toLedgerTxOut commitOutput <|)
+        & bodyTxL . reqSignerHashesTxBodyL <>~ Set.singleton (toLedgerKeyHash vkh)
+        & bodyTxL . auxDataHashTxBodyL .~ SJust (hashAlonzoTxAuxData txAuxMetadata)
+        & auxDataTxL .~ SJust txAuxMetadata
+    existingWits = ledgerBlueprintTx ^. witsTxL
+    currentInputs = ledgerBlueprintTx ^. bodyTxL . inputsTxBodyL
+    commitInputs = UTxO.inputSet utxoToCommit
+    blueprintRedeemer = unRedeemers $ toLedgerTx blueprintTx ^. witsTxL . rdmrsTxWitsL
+    reAssociatedTxIns = reAssociateOldInputs currentInputs commitInputs blueprintRedeemer
+    wits =
+      existingWits
+        & rdmrsTxWitsL .~ rebuildRedeemers reAssociatedTxIns (initialInput, initialRedeemer)
```

Comment:
can we extract all this into a lens-like function so that it can be appended with `&` to the `toLedgerTx blueprintTx` without having to extract a tmp variable?

Reply:
What do you mean? Something like:
```
    wits =
      witsTxL
        .~ ( existingWits
              & rdmrsTxWitsL .~ rebuildRedeemers reAssociatedTxIns (initialInput, initialRedeemer)
           )
   in
    fromLedgerTx $ ledgerBlueprintTx & wits

```